File: microservicio-orquestador/src/main/java/com/distribuidos/orquestador/circuit/CircuitBreaker.java
package com.distribuidos.orquestador.circuit;

import com.distribuidos.orquestador.database.DatabaseManager;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

/**
 * Implementación del Patrón de Tolerancia a Fallos Circuit Breaker.
 * Funciona como una Máquina de Estados Finitos con 3 estados:
 * 
 * - CLOSED (Cerrado): El sistema opera normalmente. Las solicitudes fluyen hacia la red.
 * - OPEN (Abierto): Se detectaron 3 fallos consecutivos. Corta la comunicación de red y devuelve respuestas Fallback inmediatas.
 * - HALF_OPEN (Semi-Abierto): Transcurridos 10 segundos de cooldown, prueba una solicitud para evaluar la recuperación autónoma del cluster.
 */
@Component
public class CircuitBreaker {

    private final String serviceName = "ServicioOrquestador";
    private final int failureThreshold = 3;        // Umbral de 3 fallos consecutivos para abrir el circuito
    private final long recoveryTimeoutMs = 10000; // 10 segundos de enfriamiento (cooldown)

    private String state = "CLOSED"; // Estado inicial del circuito
    private int failureCount = 0;    // Contador acumulado de fallos consecutivos
    private int successCount = 0;    // Contador acumulado de éxitos
    private long lastStateChange = System.currentTimeMillis(); // Timestamp del último cambio de estado

    @Autowired
    private DatabaseManager databaseManager;

    /**
     * Realiza la transición de estado de forma thread-safe y registra el evento en SQLite (circuit_log).
     */
    private synchronized void transitionTo(String newState, String reason) {
        String oldState = this.state;
        if (!oldState.equals(newState)) {
            this.state = newState;
            this.lastStateChange = System.currentTimeMillis();
            
            if ("CLOSED".equals(newState)) {
                this.failureCount = 0;
            }

            // Persiste el cambio de estado en la tabla circuit_log de SQLite
            if (databaseManager != null) {
                databaseManager.logCircuitTransition(serviceName, oldState, newState, reason);
            }
        }
    }

    /**
     * Evalúa si una solicitud puede ejecutarse o si debe bloquearse.
     * Si el circuito está OPEN y han pasado más de 10s, conmuta automáticamente a HALF_OPEN.
     */
    public synchronized boolean canExecute() {
        if ("OPEN".equals(state)) {
            long elapsed = System.currentTimeMillis() - lastStateChange;
            if (elapsed >= recoveryTimeoutMs) {
                transitionTo("HALF_OPEN", "Pasaron 10s en OPEN. Probando recuperación (HALF_OPEN).");
                return true;
            }
            return false;
        }
        return true;
    }

    /**
     * Registra una respuesta exitosa.
     * Si el circuito estaba en HALF_OPEN, se restablece a CLOSED.
     */
    public synchronized void recordSuccess() {
        this.successCount++;
        if ("HALF_OPEN".equals(state)) {
            transitionTo("CLOSED", "Prueba en HALF_OPEN exitosa. Circuito restablecido a CLOSED.");
        } else {
            this.failureCount = 0;
        }
    }

    /**
     * Registra una falla o timeout en la llamada.
     * Si se alcanza el umbral de 3 fallos en CLOSED, abre el circuito a OPEN.
     */
    public synchronized void recordFailure(String errorMsg) {
        this.failureCount++;
        if ("HALF_OPEN".equals(state)) {
            transitionTo("OPEN", "Prueba en HALF_OPEN falló (" + errorMsg + "). Circuito regresa a OPEN.");
        } else if ("CLOSED".equals(state) && failureCount >= failureThreshold) {
            transitionTo("OPEN", "Se alcanzó el umbral de " + failureThreshold + " fallos consecutivos. Circuito ABIERTO.");
        }
    }

    /**
     * Encapsula la ejecución de una función de red dentro del Circuit Breaker.
     * Lanza CircuitBreakerOpenException si el circuito se encuentra ABIERTO.
     */
    public <T> T execute(Supplier<T> supplier) {
        if (!canExecute()) {
            throw new CircuitBreakerOpenException("Circuito ABIERTO (OPEN). Petición bloqueada para evitar fallos en cascada.");
        }

        try {
            T result = supplier.get();
            recordSuccess();
            return result;
        } catch (Exception e) {
            recordFailure(e.getMessage());
            throw e;
        }
    }

    public synchronized String getState() {
        return state;
    }

    /**
     * Retorna el mapa de estado actual con el tiempo restante de cooldown en segundos.
     */
    public synchronized Map<String, Object> getStatus() {
        Map<String, Object> status = new HashMap<>();
        long elapsed = System.currentTimeMillis() - lastStateChange;
        long timeRemaining = "OPEN".equals(state) ? Math.max(0, (recoveryTimeoutMs - elapsed) / 1000) : 0;

        status.put("servicio", serviceName);
        status.put("estado", state);
        status.put("fallosConsecutivos", failureCount);
        status.put("umbralFallos", failureThreshold);
        status.put("cooldownSegundos", recoveryTimeoutMs / 1000);
        status.put("tiempoRestanteCooldownSegundos", timeRemaining);
        
        return status;
    }
}

File: orquestador/circuit_breaker.py
# orquestador/circuit_breaker.py
import time
import threading
from datetime import datetime

class CircuitBreakerOpenException(Exception):
    """Excepción lanzada cuando el Circuit Breaker está en estado OPEN."""
    def __init__(self, message="El circuito está ABIERTO (OPEN). Petición bloqueada para evitar fallos en cascada."):
        super().__init__(message)

class CircuitBreaker:
    """
    Implementación del Patrón Circuit Breaker.
    Administra los estados:
      - CLOSED    : Operación normal, monitorea fallos.
      - OPEN      : Circuito abierto por exceso de fallos, bloquea peticiones de inmediato (< 0.01s).
      - HALF_OPEN : Estado de prueba tras transcurrir recovery_timeout.
    """
    
    def __init__(self, service_name="ServicioOrquestador", failure_threshold=3, recovery_timeout=10.0, db_manager=None):
        self.service_name = service_name
        self.failure_threshold = failure_threshold  # Umbral de fallos consecutivos (3)
        self.recovery_timeout = recovery_timeout     # Tiempo de enfriamiento en segundos (10s)
        self.db_manager = db_manager

        self.state = "CLOSED"
        self.failure_count = 0
        self.success_count = 0
        self.total_calls = 0
        self.total_blocked = 0
        self.last_state_change = time.time()
        self.lock = threading.Lock()

    def _transition_to(self, new_state, reason=""):
        old_state = self.state
        if old_state != new_state:
            self.state = new_state
            self.last_state_change = time.time()
            if self.state == "CLOSED":
                self.failure_count = 0
            
            # Persistir cambio de estado en SQLite (tabla circuit_log)
            if self.db_manager:
                try:
                    self.db_manager.log_circuit_transition(
                        servicio=self.service_name,
                        estado_anterior=old_state,
                        nuevo_estado=new_state,
                        motivo=reason
                    )
                except Exception as e:
                    print(f"[CircuitBreaker DB Error] {e}")

    def can_execute(self):
        """Verifica si la petición puede ejecutarse según el estado del circuito."""
        with self.lock:
            if self.state == "OPEN":
                elapsed = time.time() - self.last_state_change
                if elapsed >= self.recovery_timeout:
                    self._transition_to("HALF_OPEN", f"Pasaron {elapsed:.1f}s en OPEN. Probando recuperación (HALF_OPEN).")
                    return True
                else:
                    self.total_blocked += 1
                    return False
            return True

    def record_success(self):
        """Registra una ejecución exitosa."""
        with self.lock:
            self.success_count += 1
            if self.state == "HALF_OPEN":
                self._transition_to("CLOSED", "Prueba en HALF_OPEN exitosa. Circuito restablecido a CLOSED.")
            else:
                self.failure_count = 0

    def record_failure(self, error_msg=""):
        """Registra un fallo en la llamada al backend."""
        with self.lock:
            self.failure_count += 1
            if self.state == "HALF_OPEN":
                self._transition_to("OPEN", f"Prueba en HALF_OPEN falló ({error_msg}). Circuito regresa a OPEN.")
            elif self.state == "CLOSED" and self.failure_count >= self.failure_threshold:
                self._transition_to("OPEN", f"Se alcanzó el umbral de {self.failure_threshold} fallos consecutivos. Circuito ABIERTO.")

    def call(self, func, *args, **kwargs):
        """
        Ejecuta la función decorada o envuelta respetando las reglas del Circuit Breaker.
        """
        with self.lock:
            self.total_calls += 1

        if not self.can_execute():
            raise CircuitBreakerOpenException()

        try:
            result = func(*args, **kwargs)
            self.record_success()
            return result
        except Exception as e:
            self.record_failure(str(e))
            raise e

    def get_status(self):
        with self.lock:
            elapsed = time.time() - self.last_state_change
            time_remaining = max(0.0, self.recovery_timeout - elapsed) if self.state == "OPEN" else 0.0
            return {
                "servicio": self.service_name,
                "estado": self.state,
                "fallos_consecutivos": self.failure_count,
                "umbral_fallos": self.failure_threshold,
                "tiempo_cooldown_seg": self.recovery_timeout,
                "tiempo_restante_cooldown_seg": round(time_remaining, 1),
                "total_llamadas": self.total_calls,
                "total_bloqueadas": self.total_blocked,
                "ultima_transicion": datetime.fromtimestamp(self.last_state_change).strftime('%Y-%m-%d %H:%M:%S')
            }
